new_audit: mixed-content https upgradeability

christhompson opened the pull request:
This introduces a new gatherer and audit that tests which requests a page makes can be directly upgraded to use HTTPS, and which fail when upgraded. The report gives a list of resources which were loaded insecurely in the default pass, their initiating document, and whether they are upgradeable. This implements the audit proposed in #3164. 

It is currently implemented as a separate custom config, as it requires an additional pass in which to do the upgrading. The new pass uses the mixed-content gatherer, which intercepts each HTTP request, and sends an HTTP 302 Found redirect to upgrade it to HTTPS. The audit compares the network records for the default pass with the upgraded requests form the mixed-content pass.

I've also included some basic tests for the functionality of both the gatherer and the audit.

This can currently be run using the following invocation:
`lighthouse-cli/index.js --chrome-flags="--headless" --config-path=./lighthouse-core/config/mixed-content.js http://www.bbc.com/`

The gatherer relies on the [Network request interception commands](https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-setRequestInterception), which may require Canary at the moment since they're currently Experimental.

The audit report looks like this (the "Full URL" column uses Lighthouse's URL type, which reveals the full URL on hover):

![audit-example](https://user-images.githubusercontent.com/113368/33406963-13c6dc5c-d524-11e7-9328-984ac25281eb.png)

christhompson commented:
One note is that the audit currently has some potential false-negatives, mostly around requests that involve redirects, where it will say a resource is non-upgradeable when it may in fact be upgradeable to HTTPS.

These tend to be most prevalent in what I sometimes refer to as "second-order requests" -- requests made by a subresource loaded by the main page (often advertising libraries). One approach I have tried is to only check and report resources loaded by the main frame (where the initiator is the same as the URL loaded by the gatherer). It's a bit of a tradeoff of complete details about the resources requested and accuracy -- if anyone feels strongly about false-negatives, I can make that change.

adrifelt commented:
Two ideas about false negatives:

1) You could put together a whitelist of common 3P inclusions with known HTTPS support and shortcut the check for those inclusions. We'd save the time for doing the second pass for that resource, and also would be able to avoid a false positive.

2) We could change the UI so that we emphasize the "Yes" resources and never clearly say "No."  The feature would be positioned as helping you find the things that are easy to switch and point out the things that are *potentially* problematic. For the "No" cases, we could say "Maybe," "Needs Investigation," or something smarter that a UX person could think of. :)

estark37 commented:
@christhompson is this waiting on review/advice from a Lighthouse person? @paulirish could you suggest someone to review? Thanks!

christhompson commented:
Yes, this should be ready for review.

Also pinging @vinamratasingal who may be able to give more feedback about the UI of the audit report.

Re: False negatives, the whitelist would be a nice addition, but there are still some cases that fall outside that (where the current implementation fails to detect a successful upgrade). We still need to load the entire page a second time, so I don't think we'd be able to short-circuit. I think the "Needs Investigation" change may be the best, especially since it frames the audit more positively overall (maybe combined with a tweak to the main audit explanation text).

vinamratasingal commented:
Hi, just got the chance to look at this. Some quick feedback (also apologies if this had been decided/discussed on a previous thread, I'm just jumping in): 

- My understanding that this was going to show up in the Best Practices section of the report. Is that still the plan or is the plan to put in its own section of the report? 
- It would be great if we could condense the top level description of the category and add some documentation explaining this. Usually we have documentation on developers.google.com for all the audits, not sure if this is something your team has already done/is planning on doing.  Suggestion on wording for the description: "These audits check whether mixed content warnings are blocking your page to upgrade to HTTPS. [Learn more]"
- Can we separate out the upgradable/insecure requests? I believe we actually show insecure requests (see screenshot below) and the specific URLs if the website is not on HTTPS, so perhaps we just list the requests that we know for sure are upgradable. 
- Is there a way that we can help users prioritize/organize the different URLs that we show (perhaps ones that are more important to upgrade than others?) 
- Also, is this audit only applicable for websites already supporting HTTPS? (i.e. if I as a dev haven't even begun migrating my site to HTTPS, this audit isn't really applicable since I need to do that before handling mixed-content warnings?) 

![screen shot 2017-12-08 at 12 24 39 pm](https://user-images.githubusercontent.com/1828723/33749755-ec2b0024-dc12-11e7-96d3-8924d9eb78d1.png)

christhompson commented:
> My understanding that this was going to show up in the Best Practices section of the report. Is that still the plan or is the plan to put in its own section of the report?

I'd be happy if we could include it in the Best Practices section. I'd thought it was not going to be included in the default config at the moment since it causes an extra page load pass.

> It would be great if we could condense the top level description of the category and add some documentation explaining this. Usually we have documentation on developers.google.com for all the audits, not sure if this is something your team has already done/is planning on doing. Suggestion on wording for the description: "These audits check whether mixed content warnings are blocking your page to upgrade to HTTPS. [Learn more]"

I can reduce the text at the top of the audit. I am planning on writing documentation that can be linked to, but was unsure of the process/order for doing that.

> Can we separate out the upgradable/insecure requests? I believe we actually show insecure requests (see screenshot below) and the specific URLs if the website is not on HTTPS, so perhaps we just list the requests that we know for sure are upgradable.

This may also help resolve the false-negatives issue. The audit could be framed as "there is no harm in upgrading these resources to HTTPS right now, to maximize your readiness to switch to HTTPS in the future".

> Is there a way that we can help users prioritize/organize the different URLs that we show (perhaps ones that are more important to upgrade than others?)

I included the Initiator column to try to add a priority level of some sort -- things initiated by the main page are more under the control of the developer, and so they would be higher priority. Maybe it would make sense to only include the resources that were initiated by the main page, especially given the discussion on the previous bullet.

> Also, is this audit only applicable for websites already supporting HTTPS? (i.e. if I as a dev haven't even begun migrating my site to HTTPS, this audit isn't really applicable since I need to do that before handling mixed-content warnings?)

This audit is intended for sites that have not yet migrated to HTTPS. The idea is that they can check whether they can just "flip the switch" to HTTPS by simply changing the relevant URLs to HTTPS, or if they need to contact specific third-parties about how to upgrade. If all URLs are upgradeable, then the developer knows they should have minimal problems when making the switch.

vinamratasingal commented:
re: Best Practices- ah, I got confused. No, the plan currently is only to expose it via the full config on the CLI, not via Best Practices. 

re: documentation- we have a tech writer (Kayce B), and it would be good to reach out to him to get the documentation written. 

re: changes- awesome. Do you mind uploading an updated screenshot so I can take another look?

christhompson commented on lighthouse-core/audits/mixed-content.js:
```diff
@@ -0,0 +1,177 @@
+      const upgradePassSecureHosts = new Set();
+      upgradedRecords.forEach(record => {
+        upgradePassHosts.add(new URL(record.url).hostname);
+        if (MixedContent.isSecureRecord(record) && record.finished) {
```
Good call. I added another check for `!record.failed`.

christhompson commented on lighthouse-core/audits/mixed-content.js:
```diff
@@ -0,0 +1,177 @@
+      insecureUrls.forEach(record => {
+        const resource = {
+          host: new URL(record.url).hostname,
+          full: record.url,
```
Done.

christhompson commented on lighthouse-core/gather/gatherers/mixed-content.js:
```diff
@@ -0,0 +1,112 @@
+ * @return {string}
+ */
+const btoa = function(str) {
+  return Buffer.from(str, 'utf8').toString('base64');
```
Yeah, that makes sense. I've pulled this inline here.

christhompson commented on lighthouse-core/test/gather/gatherers/mixed-content-test.js:
```diff
@@ -0,0 +1,117 @@
+  return Buffer.from(str, 'base64').toString('utf8');
+};
+
+describe('btoa', () =>{
```
Now that I just inlined the b64 function in the other file, I did the same here and got rid of the test. In hindsight don't think it's a particularly useful path to test (and some of the other tests should break if something strange happens to it).

christhompson commented:
Thanks for the comments. I've updated the PR to address the comments so far.

I've simplified the audit report quite a bit by restricting it to list resources that are upgradeable that are initiated by the main page only. Here's what it looks like currently:

![audit-example-2](https://user-images.githubusercontent.com/113368/33969107-53e938e6-e020-11e7-88aa-660d5e210094.png)

vinamratasingal commented:
This looks great, thanks for working through with us Chris! One last question before giving the LGTM on the UX: 

- Why are we talking about the number of insecure requests total? Should we just focus on the number of upgradable requests instead?

vinamratasingal commented:
New strings LGTM. Mind publishing an updated screenshot so I can take a final pass and LGTM?

christhompson commented:
Here's a screenshot of the latest version of the report:

![audit-example3](https://user-images.githubusercontent.com/113368/34233990-377f6b94-e59d-11e7-929c-c7ecc91273f5.png)

vinamratasingal commented:
LGTM

christhompson commented:
Pinging this thread. @paulirish As you mentioned, this currently only works with the Chrome headless flag. Would it still be possible to land this since it isn't included in any default config/category? This would then mostly be a complication to the documentation for how to run the audit.

paulirish commented:
@christhompson yup we can definitely land this.  I was playing with this audit today and had a few ideas. Instead of leaving review comments I went ahead and made the changes: https://github.com/christhompson/lighthouse/pull/2   
WDYT?

Also now that the Table and URL renderers suck less, I don't think we need the extra Hostname column in the report. Does that work for you? Or would you rather keep it explicit so they're easier to scan. I could go either way.

![image](https://user-images.githubusercontent.com/39191/34856902-c090f3ba-f6fc-11e7-84dc-56941a9c51f8.png)


If this all works, for you, lets land it. Then we can write docs on 1) how to run this audit and 2) what the audit is about.

Cheers and thanks for your patience on this one! really appreciate it.

christhompson commented:
I like the cleanliness of the new column style in your changes -- it also fixes the minor annoyance of the left/right aligned columns in the old table. Adding the `is-on-http` audit to the config is also a good idea -- thanks for adding that.

paulirish commented:
Landed! 🎉 🎉 

___

Some next steps:

* [ ] readme docs on how to run this audit
* [ ] audit docs for what this audit is about
* [ ] an npm script to make running this easier?  e.g. `yarn run mixedcontent http://www.bbc.com`

vinamratasingal commented:
Chris- are you going to take on the work of the docs? Or do we need help from @kaycebasques here? (Kayce does all of our tech docs)

christhompson commented:
I have a draft of some docs for this -- I'd be happy to coordinate with Kayce to get them in proper shape.

vinamratasingal commented:
Sounds good- let me know if you need help with this. Thanks for your patience Chris! :)
